mms/protocol/otf_message_handler.py

```diff
@@ -0,0 +1,241 @@
+            msg['requestBatch'] = list()
+            self._retrieve_request_batch(conn, msg['requestBatch'])
+
+        return "predict", msg
+
+    def retrieve_msg(self, conn):
+        """
+        Retrieve a message from the socket channel.
+
+        :param conn:
+        :return:
+        """
+        # Validate its beginning of a message
+        version = self._retrieve_double(conn)
+        if version != 1.0:
+            raise MMSError(Err.DECODE_FAILED, "Invalid message received")
+
+        cmd = self._retrieve_int(conn)
+
+        if cmd == 1:
+            cmd, msg = self._retrieve_load_msg(conn)
+        elif cmd == 2:
+            cmd, msg = self._retrieve_inference_msg(conn)
+        else:
+            return "unknown", "Wrong command"
+
+        eom = self._retrieve_buffer(conn, 2)
+
+        if eom != b'\r\n':
+            raise MMSError(Err.DECODE_FAILED, "Invalid message received")
+
+        return cmd, msg
+
+    def _encode_inference_response(self, kwargs):
+        try:
+            req_id_map = kwargs['req_id_map']
+            invalid_reqs = kwargs['invalid_reqs']
+            ret = kwargs['resp']
+            msg = bytearray()
+            msg += struct.pack('!i', -1)  # start of list
+            content_type = ""
+            for idx, val in enumerate(ret):
+                msg += struct.pack("!i", len(req_id_map[idx]))
+                msg += struct.pack('!{}s'.format(len(req_id_map[idx])), req_id_map[idx].encode())
+
+                # msg += struct.pack('!i', 200) # CODE
+
+                if isinstance(val, str):
+                    content_type = "text"
+                    msg += struct.pack('!i', len(content_type))
+                    msg += struct.pack('!{}s'.format(len(content_type)), content_type.encode())
+
+                    msg += struct.pack('!i', len(val))
+                    msg += struct.pack('!{}s'.format(len(val)), val.encode())
+                elif isinstance(val, bytes):
+                    content_type = "binary"
+                    msg += struct.pack('!i', len(content_type))
+                    msg += struct.pack('!{}s'.format(len(content_type)), content_type.encode())
+
+                    msg += struct.pack('!i', len(val))
+                    msg += val
+                else:
+                    content_type = 'json'
+                    json_value = json.dumps(val)
+                    msg += struct.pack('!i', len(content_type))
+                    msg += struct.pack('!{}s'.format(len(content_type)), content_type.encode())
+
+                    msg += struct.pack('!i', len(json_value))
+                    msg += struct.pack('!{}s'.format(len(json_value)), json_value.encode())
+
+            for req in invalid_reqs.keys():
+                msg += struct.pack('!i', len(req))
+                msg += struct.pack('!{}s'.format(len(req)), req.encode())
+
+                msg += struct.pack('!i', 4)
+                msg += struct.pack('!i', invalid_reqs.get(req))
```

Comment:
Only use .get if you want a default value instead of a keyError. Otherwise, use the nicer invalid_reqs[req]

Reply:
Yup. Thats the idea. To get a None or value and not jsut get a keyerror.